Mixing/Base/src/PileupRandomNumberGenerator.cc

```diff
@@ -0,0 +1,101 @@
+#include "CLHEP/Random/RanecuEngine.h"
+#include "CLHEP/Random/MixMaxRng.h"
+#include "IOMC/RandomEngine/interface/TRandomAdaptor.h"
+
+CLHEP::HepRandomEngine& PileupRandomNumberGenerator::getEngine(edm::StreamID const&) {
+  return *m_modulesToEngines[findPresentModule()];
+}
+
+CLHEP::HepRandomEngine& PileupRandomNumberGenerator::getEngine(edm::LuminosityBlockIndex const&) {
+  return *m_modulesToEngines[findPresentModule()];
+}
+
+std::unique_ptr<CLHEP::HepRandomEngine> PileupRandomNumberGenerator::cloneEngine(edm::LuminosityBlockIndex const&) {
+  auto& engine = m_modulesToEngines[findPresentModule()];
+  return cloneEngine(*engine);
+}
+
+void PileupRandomNumberGenerator::setSeed(uint32_t iSeed) { m_seed = iSeed; }
+
+void PileupRandomNumberGenerator::setEngine(CLHEP::HepRandomEngine const& iEngine) {
+  for (auto& v : m_modulesToEngines) {
+    v.second = cloneEngine(iEngine);
+  }
+}
+
+void PileupRandomNumberGenerator::preBeginLumi(edm::LuminosityBlock const& lumi) {}
+void PileupRandomNumberGenerator::postEventRead(edm::Event const& event) {}
+
+void PileupRandomNumberGenerator::setLumiCache(edm::LuminosityBlockIndex,
+                                               std::vector<RandomEngineState> const& iStates) {}
+void PileupRandomNumberGenerator::setEventCache(edm::StreamID, std::vector<RandomEngineState> const& iStates) {}
+
+std::vector<RandomEngineState> const& PileupRandomNumberGenerator::getEventCache(edm::StreamID const&) const {
+  static const std::vector<RandomEngineState> s_dummy;
+  return s_dummy;
+}
+std::vector<RandomEngineState> const& PileupRandomNumberGenerator::getLumiCache(edm::LuminosityBlockIndex const&) const {
+  static const std::vector<RandomEngineState> s_dummy;
+  return s_dummy;
+}
+
+void PileupRandomNumberGenerator::consumes(edm::ConsumesCollector&& iC) const {}
+
+void PileupRandomNumberGenerator::print(std::ostream& os) const {}
+
+std::unique_ptr<CLHEP::HepRandomEngine> PileupRandomNumberGenerator::cloneEngine(
+    CLHEP::HepRandomEngine const& existingEngine) {
+  std::vector<unsigned long> stateL = existingEngine.put();
+  long seedL = existingEngine.getSeed();
+  std::unique_ptr<CLHEP::HepRandomEngine> newEngine;
+  if (stateL[0] == CLHEP::engineIDulong<CLHEP::HepJamesRandom>()) {
+    newEngine = std::make_unique<CLHEP::HepJamesRandom>(seedL);
+  } else if (stateL[0] == CLHEP::engineIDulong<CLHEP::RanecuEngine>()) {
+    newEngine = std::make_unique<CLHEP::RanecuEngine>();
+  } else if (stateL[0] == CLHEP::engineIDulong<CLHEP::MixMaxRng>()) {
+    newEngine = std::make_unique<CLHEP::MixMaxRng>(seedL);
+  } else if (stateL[0] == CLHEP::engineIDulong<edm::TRandomAdaptor>()) {
+    newEngine = std::make_unique<edm::TRandomAdaptor>(seedL);
+  } else {
+    // Sanity check, it should not be possible for this to happen.
+    throw edm::Exception(edm::errors::Unknown)
+        << "The RandomNumberGeneratorService is trying to clone unknown engine type\n";
+  }
+  newEngine->get(stateL);
```

Comment:
The code looks very similar to `RandomNumberGeneratorService::cloneEngine()`. Would it be worth of abstracting that into a separate function? (even if we don't add support for new random number generators that often)

Reply:
done